Routing driver forgets addresses on some errors

lutovich opened the pull request:
This PR makes routing driver forget addresses and purge their connections on network and database errors. It also makes driver forget write server addresses when they fail as writers, connections remain in the pool because such machines can remain readers/routers.

lutovich committed:
Routing driver forgets addresses on some errors

This commit makes routing driver forget addresses and purge their
connections on network and database errors. It also makes driver
forget write server addresses when they fail as writers, connections
remain in the pool because such machines can remain readers/routers.

zhenlineo commented:
LGTM. CC @technige

technige reviewed (changes requested):
Not keen on the `error_handler` changes to `connection.py`. It's quite unusual to dynamically attach methods in this way and the new `try...except` blocks catch all `Exception` types, which is generally considered too brute force.

I'm happy to pair on this.

lutovich committed:
Add new more specific error types

To avoid comparing strings when handling routing errors.

lutovich committed:
Improve connection error handling infra

Connection now takes a dedicated error handler object in constructor
instead of having one dynamically attached after creation.

lutovich commented:
@technige dynamic `error_handler` is now replaced by constructor injection of `ConnectionErrorHandler`. Later is different for direct and routing driver.
